[TensorRT EP] Load precompiled TRT engine file directly

chilo-ms opened the pull request:
When the TRT engine cache (precompiled engine) is present, it doesn't make sense to go over the processes of model verification, model optimization, TRT EP's GetCapability(), TRT EP's model proto reconstruction, calling TRT parser and engine compilation.
This PR makes TRT EP skip those processes and directly load the engine to perform inference.

The feature request: https://github.com/microsoft/onnxruntime/issues/18072

Features:

- Replace original model with TRT engine wrapped ONNX model. It can save a lot of time as mentioned above.

- How to get TRT engine wrapped ONNX model?
    1. Set `trt_dump_ep_context_model` provider option to "true" and run the inference. You will find the "xxx_wrapper.onnx" at the engine cache path. (The same logic of generating engine cache)
    2. Use gen_trt_engine_wrapper_onnx_model.py

- Three provider options are added, 
    `trt_dump_ep_context_model`: Enable dump wrapped onnx model by TRT EP
    `trt_ep_context_embed_mode`: Add embed_mode as attribute. 0 means engine cache path, 1 means engine binary data.
    `trt_ep_context_compute_capability_enable`: Add hardware_arch as attribute. When running the model, TRT EP will check consistency between model's hardware_arch and GPU's compute capability.

- When the engine cache path is given in the wrapped model, TRT EP will first search for the engine file using the path (relative to model path), if it can't find it, it will change to use the path as it is (depends on user, could be relative to working dir or absolute path)

Note: 

1. This PR includes the change of https://github.com/microsoft/onnxruntime/pull/17751


Constraints:

1. The whole model should be fully supported by TRT. 
4. Users need to make sure the engine is built with min/max/opt optimization profiles that large enough to cover the range of all inputs. TRT EP will simply fail and won't rebuild the engine if the input shape is out of range during runtime.

gedoensmax commented on onnxruntime/test/python/onnxruntime_test_engine_wrapper.py:
```diff
@@ -0,0 +1,99 @@
+
+        # Second run to test ctx node with engine cache path
+        self.create_ctx_node(cache_path=os.path.join(self.trt_engine_cache_path_, cache_name))
+        providers = [("TensorrtExecutionProvider", {})]
```
If you ask me it is kind of misleading that there is an input variable `providers` that is overridden here. I would move them both in this functions scope to improve readability.

chilo-ms replied:
Refactor the unit test to make it more readable.

gedoensmax reviewed (commented):
I have only quickly read through the interfacing python code which is easy to understand  and looks like what we discussed. If the PR is still open end of next week I will give it a proper review and see if any requirements are missing from my side.

jywu-msft commented:
> I have only quickly read through the interfacing python code which is easy to understand and looks like what we discussed. If the PR is still open end of next week I will give it a proper review and see if any requirements are missing from my side.

yes, we would like to get yours and others' feedback before merging this PR

chilo-ms commented:
I'm still working on writing out the engine wrapped onnx model firectly from TRT EP, so that users don't need to run the script offline to generate the onnx model.

gedoensmax commented on onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:
```diff
@@ -689,6 +693,420 @@ Status ApplyProfileShapesFromInputTensorValue(std::vector<nvinfer1::IOptimizatio
+    for (int j = 0, end = nb_dims; j < end; ++j) {
+      dims.d[j] = static_cast<int32_t>(tensor_shapes[j]);
+    }
+    if (!trt_context->setInputShape(input_name, dims)) {
+      std::string error_input_name = input_name;
+      ORT_THROW_IF_ERROR(ORT_MAKE_STATUS(ONNXRUNTIME, EP_FAIL,
```
Running this command: `onnxruntime_perf_test -e tensorrt -M -I -F "0:2" -F "2:256" -F "3:256" trt_path_embed.onnx` did not work for the model that is embedded like this:
```
['input1']
[<DataType.HALF: 1>]
[(-1, 8, -1, -1)]
['output1']
[<DataType.HALF: 1>]
[(-1, 128, -1, -1)]
```
How should I test the model with perf test ? I compiled the engine externally and the model is just 2 convolutions for testing with following min max shapes:
```
     --minShapes="input1:2x8x128x128" 
     --optShapes="input1:2x8x256x256" 
     --maxShapes="input1:4x8x512x512"
```
Perf test alway sets the dimensions to `(1,8,1,1)`

chilo-ms replied:
Good catch.

ORT interprets "-1" of dimension as empty so you can't see the effect of override free dimensions from onnxruntime_perf_test.
ORT uses symbolic dimension for dynamic dimension, so i modified the script to generate dynamic dimension with `free_dim_0` ,` free_dim_1` ...

Also, please use `-f` to specify a free dimension by name not `-F` . So please run like this:
`onnxruntime_perf_test -e tensorrt -M -I -F "free_dim_0:2" -F "free_dim_1:256" -F "free_dim_2:256" trt_path_embed.onnx`

gedoensmax replied:
Yeah I usually use `-f ` I just tried both for TRT. Using actual names sounds much more reasonable.

gedoensmax reviewed (commented):
I was very happy with the results of my testing with this. Are you still spanning on adding the option to instead of dumping the engine to disk dump it to an ONNX immediately ? If we do this we would even be able to keep the exact input nodes of the model like this 
```
    ---- 1 Graph Input(s) ----
    {input1 [dtype=float16, shape=('batch_size', 8, 'height', 'width')]}
```
instead of resulting in something like this from the generation script:
```
    ---- 1 Graph Input(s) ----
    {input1 [dtype=float16, shape=(-1, 8, -1, -1)]}
```
Also if I use the `-u` option on onnxruntime_perf test it still tells me:
```
Unable to serialize model as it contains compiled nodes. Please disable any execution providers which generate compiled nodes.
```
Wouldn't this be the preferred way to dump a model with an integrated engine ?

chilo-ms commented on onnxruntime/core/providers/tensorrt/onnx_ctx_model_helper.h:
```diff
@@ -0,0 +1,47 @@
+static const std::string EPCONTEXT_OP = "EPContext";
+static const std::string MAIN_CONTEXT = "main_context";
+static const std::string EMBED_MODE = "embed_mode";
+static const std::string EP_CACHE_CONTEXT = "ep_cache_context";
```
Considering the types of AttributeProto that ONNX supports, I'm afraid that the string is the most suitable one for the ep_cache_context.
For those types that won't be hashed in ORT, we can't use AttributeProto.GRAPH since it's not compatible with TRT engine cache. I thought about AttributeProto.UNDEFINED, but [make_attribute](https://github.com/onnx/onnx/blob/main/onnx/helper.py#L838) in ONNX doesn't handle the undefined type case. So, the script can't generate the attribute with undefined type.
Therefore, i think setting a low optimization level in session option is the better option for now.

gedoensmax replied:
Ok understood. We should add this to the docs that it is preferred to store a path to an engine rather than storing the actual engine inside the ONNX.

chilo-ms commented:
> I was very happy with the results of my testing with this. Are you still spanning on adding the option to instead of dumping the engine to disk dump it to an ONNX immediately ? 

Actually I'm not really starting to add the option to dump it to an ONNX that wraps the engine.
Hopefully i can start to do so.



> Also if I use the `-u` option on onnxruntime_perf test it still tells me:
> 
> ```
> Unable to serialize model as it contains compiled nodes. Please disable any execution providers which generate compiled nodes.
> ```
> 
> Wouldn't this be the preferred way to dump a model with an integrated engine ?

`-u` is to save optimized onnx model. But we can't do this when using TRT EP since TRT EP requires the "Compile" method to generate the customized kernel/node which is not a standard ONNX format to be serialized.

gedoensmax commented:
> -u is to save optimized onnx model which happens at the end of session initialization and before inference time.

Ok understood that `-u` is not suitable. Do you want to work on dumping the engine as an ONNX in another PR or will it still be part of this one ? I really like the idea of doing that inside the EP as we can have the same node names and free dim names as the original ONNX which reduces a lot of friction.

chilo-ms commented:
> > -u is to save optimized onnx model which happens at the end of session initialization and before inference time.
> 
> Ok understood that `-u` is not suitable. Do you want to work on dumping the engine as an ONNX in another PR or will it still be part of this one ? I really like the idea of doing that inside the EP as we can have the same node names and free dim names as the original ONNX which reduces a lot of friction.

I started to work on this now. It seems we won't merge this PR soon due to holiday week, so i will add this feature as part of this PR.

chilo-ms commented:
@gedoensmax, I added the feature of dumping TRT engine wrapped ONNX model by TRT EP.
Please revisit the PR description for how to use the `trt_dump_ep_context_model`

gedoensmax commented:
This looks good to me with one minor adjustment - could we rename the dumped file `TensorrtExecutionProvider_TRTKernel_graph_torch_jit_14443365187856741147_0_0_sm89_wrapper.onnx` to e.g. `trt_wrapper.onnx` ? That would make the name deterministic and makes it easy to check for the existence of this file. If the engine changes the link to the engine will just be different inside the file. 
Also the link to then engine in the file is `test\TensorrtExecutionProvider_TRTKernel_graph_torch_jit_14443365187856741147_0_0_sm89.engine` when I define the cache path as `test`. But as the ONNX is already dumped to the cache folder the actual relative path should be just `TensorrtExecutionProvider_TRTKernel_graph_torch_jit_14443365187856741147_0_0_sm89.engine`.

Oh also could you take care of adding a note on the docs (whenever you add the option to the docs) that it is recommended to use optimization level 0 ? I am wondering if optimization level 0 would ne be advisable in general for TRT.

chilo-ms commented:
> This looks good to me with one minor adjustment - could we rename the dumped file `TensorrtExecutionProvider_TRTKernel_graph_torch_jit_14443365187856741147_0_0_sm89_wrapper.onnx` to e.g. `trt_wrapper.onnx` ? That would make the name deterministic and makes it easy to check for the existence of this file. If the engine changes the link to the engine will just be different inside the file. 

I'm thinking about following scenario.

What about there are more than one engine that are built against different GPU hardware, e.g. different compute capability? In this case, users need different cache folders to hold the corresponding `trt_wrapper.onnx`s. 
Also, if engine is embedded in the wrapper model as the binary data, with the name of `trt_wrapper.onnx`, users might not know what engine is embedded in the wrapper model.

chilo-ms commented:
>Oh also could you take care of adding a note on the docs (whenever you add the option to the docs) that it is recommended to use optimization level 0 ? I am wondering if optimization level 0 would ne be advisable in general for TRT.

yes, i will document that it's suggested to use optimization level 0 when embeds the engine binary data in wrapper model.

I think we should suggest to use optimization level 0 for TRT in general. But we haven't thoroughly compared the optimizations between ORT and TRT. I assume there should be a lot of overlaps, if that's the case, we can let TRT do all the optimizations.
But it still has some minor drawbacks, for example, the model proto fed to TRT will be larger.

gedoensmax commented:
Hm true the sm version we can resolve by also embedding this in the file name, but I dislike having the model hash in there. I am happy to other proposals though. 
Ideally there would be some API that can be used to return the ONNX maybe ? Or we make the compute capability a list. In case it is a list we select the correct engine somehow ? 

Following scenarios: 
- I compile, but on the second call I want to reuse the embedded ONNX as it loads much quicker.
- I have compiled engines for multiple sm versions and want to use the correct one given the current device. 

In any way I want to have a definitive way of getting the correct ONNX for the first I would really prefer a deterministic way or an API how I can query the ONNX name. 
For the second we have to make the decision if the user should query the device and pass in a correct ONNX or If one ONNX can hold multiple compute capabilities. Let me know what you think.

gedoensmax commented:
@jywu-msft as discussed in our last sync the idea was to quick load an engine if the engine path is not a folder, but a file. Do we plan to include this here or should this be another PR ?

jywu-msft commented:
> @jywu-msft as discussed in our last sync the idea was to quick load an engine if the engine path is not a folder, but a file. Do we plan to include this here or should this be another PR ?

i think we can probably add it on as another PR.  
the plan is to get https://github.com/microsoft/onnxruntime/pull/17751 merged first,
and then this PR, then follow-up PR's can be made to further enhance it.

gedoensmax commented:
@jywu-msft Since `enqueuev3` is now what are the plans for this PR ?  Should I work on a rebase or do we let it rest until Chi returns ?

jywu-msft commented:
> @jywu-msft Since `enqueuev3` is now what are the plans for this PR ? Should I work on a rebase or do we let it rest until Chi returns ?

plan is to merge this by end of week. @chilo-ms will work on it.

jywu-msft commented on include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h:
```diff
@@ -46,4 +46,7 @@ struct OrtTensorRTProviderOptionsV2 {
   const char* trt_profile_max_shapes{nullptr};           // Specify the range of the input shapes to build the engine with
   const char* trt_profile_opt_shapes{nullptr};           // Specify the range of the input shapes to build the engine with
   int trt_cuda_graph_enable{0};                          // Enable CUDA graph in ORT TRT
+  int trt_dump_ep_context_model{0};                      // Dump EP context node model
```
fyi https://github.com/microsoft/onnxruntime/pull/18877
added these options at session level. so all EP's can use them.

chilo-ms replied:
After taking a closer look, i think it requires more changes for TRT EP to consume session options.
QNN EP factory [creation](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/session/provider_registration.cc#L97) is in `OrtApis::SessionOptionsAppendExecutionProvider` where it passes session options to QNN EP factory constructor.
However, the current TRT EP factory [initialization](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/session/provider_bridge_ort.cc#L1696) does not consume session options. However, we can still make it work but requires more work that is not related to this "direct load" PR.

So, i kind of want to keep those TRT EP provider options as an option. We can have another PR to make TRT EP consume session options later.

chilo-ms replied:
> ok, let's keep it separate for now. I just want to confirm the session option support can be done in time for ort 1.17. we don't want the situation where 1.17 supports trt ep option and then session option after the release.
It can be done for ORT 1.17 and i'm starting to work on this now

chilo-ms commented:
@jywu-msft @gedoensmax

Github isn't smart enough to make this PR smoothly merge main. So, i have to resolve a lot of conflicts ... 

The key change regarding the code path inside `TensorrtExecutionProvider::Compile()`is I added two functions: `CreateNodeComputeInfoFromPrecompiledEngine()` and `CreateNodeComputeInfoFromGraph()`
I moved almost all the code that originally was in Compile() into `CreateNodeComputeInfoFromGraph()`.
For `CreateNodeComputeInfoFromPrecompiledEngine()`, it's the streamlined function that only keeps essential setting for directly loading the engine.
Hope this explanation help review.

gedoensmax commented on onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:
```diff
@@ -2280,6 +2300,19 @@ bool TensorrtExecutionProvider::DetectTensorRTGraphCycles(SubGraphCollection_t&
+  // If the model consists of only a single "EPContext" contrib op, it means TRT EP can fetch the precompiled engine info from the node and
+  // load the engine directly without having to go through the processes of graph proto reconstruction, calling TRT parser and engine compilation.
+  // So, simply return the ComputeCapability here.
+  if (graph.NumberOfNodes() == 1 && GraphHasCtxNode(graph)) {
+    SubGraph_t supported_node_vector = {{0}, true};
```
Wouldn't the condition to bypass the builder rather be that all nodes are EP context nodes ? In theory we could have an ONNX file with multiple EP Context nodes. It's probably not possible currently but it could even be multiple engines connected with control flow nodes that are executed on CPU ?

jywu-msft replied:
yes, we are in the process of designing multiple context node support. (this is needed for other EP's as well)
For now, can we explicitly enforce that we can only handle a single EP context node here?
If there is more than one, we need to throw exception?

gedoensmax commented on onnxruntime/core/providers/tensorrt/onnx_ctx_model_helper.cc:
```diff
@@ -0,0 +1,246 @@
+ * The sanity check for EP context contrib op.
+ */
+bool TensorRTCacheModelHandler::ValidateEPCtxNode(const GraphViewer& graph_viewer) {
+  assert(graph_viewer.NumberOfNodes() == 1);
+  assert(graph_viewer.GetNode(0)->OpType() == EPCONTEXT_OP);
+  auto node = graph_viewer.GetNode(0);
+  auto& attrs = node->GetAttributes();
```
Could we add a warning here that it is recommended to disable graph optimization and embed the engine via its path for the best load time ?

chilo-ms replied:
Added in the follow-up PR

gedoensmax reviewed (commented):
The PR looks good to me, the few comments I left could be considered as nitpicks to be honest.

jywu-msft commented on onnxruntime/core/graph/contrib_ops/contrib_defs.cc:
```diff
@@ -3230,6 +3230,11 @@ void RegisterContribSchemas() {
           AttributeProto::STRING,
           OPTIONAL_VALUE)
+      .Attr(
+          "hardware_arch",
```
this schema change triggered kernel doc generation failure? seems like something else needs to change? maybe some doc file?

chilo-ms replied:
Updated the doc as well.

jywu-msft commented on onnxruntime/core/graph/contrib_ops/contrib_defs.cc:
```diff
@@ -3230,6 +3230,11 @@ void RegisterContribSchemas() {
           OPTIONAL_VALUE)
+      .Attr(
+          "hardware_arch",
+          "(Optional) Hardware architecture of the GPU.",
```
change to hardware specific string.  (remove mention of GPU)

chilo-ms replied:
Changed.

jywu-msft commented:
> The PR looks good to me, the few comments I left could be considered as nitpicks to be honest.

thanks. these suggestions will be added in a follow-up PR. we will merge this one asap as soon as CI's pass.

chilo-ms commented:
> @jywu-msft as discussed in our last sync the idea was to quick load an engine if the engine path is not a folder, but a file. Do we plan to include this here or should this be another PR ?

Please see the [PR ](https://github.com/microsoft/onnxruntime/pull/19154)here for naming the "EP context" model. Feel free to provide feedbacks.
